Proiect3D/11_01_umbra.cpp

```diff
@@ -413,6 +409,51 @@ void RenderFunction(void)
 				GL_UNSIGNED_SHORT,
 				(void*)(((2 * (NR_PARR + 1) * (NR_MERID)+4 * patr) + 68) * sizeof(GLushort)));
 	}
+}
+void RenderFunction(void)
+{
+	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+	glEnable(GL_DEPTH_TEST);
+
+	//pozitia observatorului
+	Obsx = Refx + dist * cos(alpha) * cos(beta);
+	Obsy = Refy + dist * cos(alpha) * sin(beta);
+	Obsz = Refz + dist * sin(alpha);
+
+	// matrice de vizualizare + proiectie
+	glm::vec3 Obs = glm::vec3(Obsx, Obsy, Obsz);   // se schimba pozitia observatorului	
+	glm::vec3 PctRef = glm::vec3(Refx, Refy, Refz); // pozitia punctului de referinta
+	glm::vec3 Vert = glm::vec3(Vx, Vy, Vz); // verticala din planul de vizualizare 
+	view = glm::lookAt(Obs, PctRef, Vert);
+	glUniformMatrix4fv(viewLocation, 1, GL_FALSE, &view[0][0]);
+	projection = glm::infinitePerspective(fov, GLfloat(width) / GLfloat(height), znear);
+	glUniformMatrix4fv(projLocation, 1, GL_FALSE, &projection[0][0]);
+
+	// matricea pentru umbra
+	float D = -10.0f;
+	matrUmbra[0][0] = zL + D; matrUmbra[0][1] = 0; matrUmbra[0][2] = 0; matrUmbra[0][3] = 0;
+	matrUmbra[1][0] = 0; matrUmbra[1][1] = zL + D; matrUmbra[1][2] = 0; matrUmbra[1][3] = 0;
+	matrUmbra[2][0] = -xL; matrUmbra[2][1] = -yL; matrUmbra[2][2] = D; matrUmbra[2][3] = -1;
+	matrUmbra[3][0] = -D * xL; matrUmbra[3][1] = -D * yL; matrUmbra[3][2] = -D * zL; matrUmbra[3][3] = zL;
+	glUniformMatrix4fv(matrUmbraLocation, 1, GL_FALSE, &matrUmbra[0][0]);
+
+	// Variabile uniforme pentru iluminare
+	glUniform3f(lightColorLocation, 1.0f, 1.0f, 1.0f);
+	glUniform3f(lightPosLocation, xL, yL, zL);
+	glUniform3f(viewPosLocation, Obsx, Obsy, Obsz);
+
+	myMatrix = glm::mat4(1.0f);
+	DrawBalloon(myMatrix,0);
+	//vector random pt x,y
+	l = l + 0.8;
+	for (int i = 1; i <= 20; i++)
+	{
+		matrTransl = glm::translate(glm::mat4(1.0f), glm::vec3(xpos[i], ypos[i], std::fmod((l + i * 120), 5000.0f)));
```

Comment:
I don't think an offset of 120 is enough, some ballons might still overlap. There shouldn't be multiple balloons on the same height level.
![image](https://github.com/andlft/OpenGL-3D-Project/assets/92049917/e05e0faf-43be-4054-9978-3d705027a6d2)